api/grpcserver/mesh_service_test.go

```diff
@@ -0,0 +1,228 @@
+	bp.Messages[1].SmesherID = sig.NodeID()
+	mp := &types.MalfeasanceProof{
+		Layer: types.LayerID(layer),
+		Proof: types.Proof{
+			Type: types.MultipleBallots,
+			Data: &bp,
+		},
+	}
+	data, err := codec.Encode(mp)
+	require.NoError(tb, err)
+	require.NoError(tb, identities.SetMalicious(db, sig.NodeID(), data, time.Now()))
+	return sig.NodeID(), mp
+}
+
+func HareMalfeasance(tb testing.TB, db sql.Executor) (types.NodeID, *types.MalfeasanceProof) {
+	sig, err := signing.NewEdSigner()
+	require.NoError(tb, err)
+	hp := types.HareProof{
+		Messages: [2]types.HareProofMsg{
+			{
+				InnerMsg: types.HareMetadata{
+					Layer:   types.LayerID(layer),
+					Round:   3,
+					MsgHash: types.RandomHash(),
+				},
+			},
+			{
+				InnerMsg: types.HareMetadata{
+					Layer:   types.LayerID(layer),
+					Round:   3,
+					MsgHash: types.RandomHash(),
+				},
+			},
+		},
+	}
+	hp.Messages[0].Signature = sig.Sign(signing.HARE, hp.Messages[0].SignedBytes())
+	hp.Messages[0].SmesherID = sig.NodeID()
+	hp.Messages[1].Signature = sig.Sign(signing.HARE, hp.Messages[1].SignedBytes())
+	hp.Messages[1].SmesherID = sig.NodeID()
+	mp := &types.MalfeasanceProof{
+		Layer: types.LayerID(layer),
+		Proof: types.Proof{
+			Type: types.HareEquivocation,
+			Data: &hp,
+		},
+	}
+	data, err := codec.Encode(mp)
+	require.NoError(tb, err)
+	require.NoError(tb, identities.SetMalicious(db, sig.NodeID(), data, time.Now()))
+	return sig.NodeID(), mp
+}
+
+func TestMeshService_MalfeasanceQuery(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	genTime := NewMockgenesisTimeAPI(ctrl)
+	db := sql.InMemory()
+	srv := NewMeshService(datastore.NewCachedDB(db, logtest.New(t)), meshAPIMock, conStateAPI, genTime, layersPerEpoch, types.Hash20{}, layerDuration, layerAvgSize, txsPerProposal, logtest.New(t).WithName("grpc.Mesh"))
+	t.Cleanup(launchServer(t, cfg, srv))
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	conn := dialGrpc(ctx, t, cfg.PublicListener)
+	client := pb.NewMeshServiceClient(conn)
+	nodeID, proof := BallotMalfeasance(t, db)
+
+	req := &pb.MalfeasanceRequest{
+		Smesher:      &pb.SmesherId{Id: nodeID.Bytes()},
+		IncludeProof: true,
+	}
+	resp, err := client.MalfeasanceQuery(context.Background(), req)
+	require.NoError(t, err)
+	require.Equal(t, nodeID, types.BytesToNodeID(resp.Proof.SmesherId.Id))
+	require.EqualValues(t, layer, resp.Proof.Layer.Number)
+	require.Equal(t, pb.MalfeasanceProof_MALFEASANCE_BALLOT, resp.Proof.Kind)
+	require.Equal(t, events.ToMalfeasancePB(nodeID, proof, true), resp.Proof)
+	require.NotEmpty(t, resp.Proof.Proof)
+	var got types.MalfeasanceProof
+	require.NoError(t, codec.Decode(resp.Proof.Proof, &got))
+	require.Equal(t, *proof, got)
+
+	req.IncludeProof = false
+	resp, err = client.MalfeasanceQuery(context.Background(), req)
+	require.NoError(t, err)
+	require.Empty(t, resp.Proof.Proof)
+}
+
+func TestMeshService_MalfeasanceStream(t *testing.T) {
```

Comment:
Would it make sense to extend this or create a new test that persists malfeasance proofs while they are streamed?

I think it would make sense to have such a test to check for races and if every proof is only streamed once?

Reply:
persisting malfeasance doesn't cause event to trigger. the code there also calls `events.ReportMalfeasance(id, proof)`, like i did in this test. i don't see there is a difference btwn what you are suggesting and the test is currently doing.